### PR TITLE
Refactor object.d: Move isStaticArray to core.internal.traits

### PR DESCRIPTION
**Pull request by JinShil**
There was an `_isStaticArray` implementation in `object.d` to support `destroy`.  However, [that functionality is already in `std.traits`](https://dlang.org/phobos/std_traits.html#isStaticArray).  So I removed the duplicate implementation in `object.d` and copied it to `core.internal.traits`.  A subsequent PR to Phobos can then be created to forward to the `core.internal.traits` implementation.

I may even follow up with a PR to deprecate [`__traits(isStaticArray)`](https://dlang.org/spec/traits.html#isStaticArray) in favor of the library implementation.

### CODE REVIEW

File: src/core/internal/traits.d [hunk 1]
```diff
@@ -164,6 +164,26 @@ template isInnerClass(T) if (is(T == class))
         enum isInnerClass = false;
 }
 
+// Detect whether type `T` is a static array.
+enum bool isStaticArray(T) = __traits(isStaticArray, T);
```
Comment (JinShil) [inline src/core/internal/traits.d]: Copied from https://github.com/dlang/phobos/blob/991331a951d68de9b36e699ac7572c1d407f7560/std/traits.d#L6646-L6663

**Comment by ZombineDev** [comment]
Replace both declarations with:
```D
enum isStaticArray(T) = is(T : U[N], U, size_t N);
```

(GitHub is acting a bit strange - it didn't include my comment in the review.)

**Comment by jacob-carlborg** [comment]
> I may even follow up with a PR to deprecate __traits(isStaticArray) in favor of the library implementation.

Do we know why `__traits(isStaticArray)` was added?

**Comment by JinShil** [comment]
> Do we know why `__traits(isStaticArray)` was added?

I don't know for sure yet.  My guess at the moment is that it was added before D had all of these fantastic introspection features.  I'll research it before submitting a PR to deprecate it (assuming I even get around to it).

**Comment by jacob-carlborg** [comment]
> Do we know why __traits(isStaticArray) was added?

Seems it was part of the first implementation of `__traits` https://github.com/dlang/dmd/commit/e56081448f and that Brad committed the whole release. Most likely before GitHub.

File: src/core/internal/traits.d [hunk 2]
```diff
@@ -164,6 +164,26 @@ template isInnerClass(T) if (is(T == class))
         enum isInnerClass = false;
 }
 
+/// Detect whether type `T` is a static array.
+enum isStaticArray(T) = is(T : U[N], U, size_t N);
```
Comment (ZombineDev) [inline src/core/internal/traits.d]: I think the only question that remains is whether we want to allow implicit conversions, i.e. types which have an `alias this` to a static array. Or `enum`s whose base type is a static array:
https://run.dlang.io/gist/run-dlang/e18c94ccff92c2d09feba048b7700860?compiler=dmd

Comment (JinShil) [reply src/core/internal/traits.d]: > I think the only question that remains is whether we want to allow implicit conversions, i.e. types which have an alias this to a static array

I say *no*.  We want to know if it *is* a static array not whether or not it implicitly converts to a static array or behaves like a static array.

> Or enums whose base type is a static array:
I say *yes*.  That is an enumeration of static arrays.  The enum items are indeed static arrays.

The implementation has been updated and test cases have been added.  Nice work identifying these edge cases!

**Comment by JinShil** [comment]
Another test case that needs adjudication:
```D
int[1] x;

pragma (msg, "isStaticArray!x: ", isStaticArray!x);  // Compiler error
pragma (msg, "__traits(isStaticArray, x): ", __traits(isStaticArray, x)); // true
```
I'd like for the `isStaticArray` template to return `true` for this case, but I'm not sure how to write it to do so.

**Comment by MoonlightSentinel** [comment]
> I'd like for the `isStaticArray` template to return `true` for this case, but I'm not sure how to write it to do so.

You could make `T` an alias parameter and obtain its type by using something like
```
template Type(alias T) {
    static if(is(typeof(T) == X, X)) {
        alias Type = X;
    }
    else {
        alias Type = T;
}
```

But that is probably out of scope for this PR.

**Comment by JinShil** [comment]
> 
> 
> Another test case that needs adjudication:
> 
> ```d
> int[1] x;
> 
> pragma (msg, "isStaticArray!x: ", isStaticArray!x);  // Compiler error
> pragma (msg, "__traits(isStaticArray, x): ", __traits(isStaticArray, x)); // true
> ```
> 
> I'd like for the `isStaticArray` template to return `true` for this case, but I'm not sure how to write it to do so.

Actually, I think the user should probably be required to type `isStaticArray!(typeof(x))`

**Comment by JinShil** [comment]
I used @MoonlightSentinel's technique and added a few more tests to keep the implementation consistent with `__traits(istStaticArray)`.  This PR is ready for additional scrutiny.

File: src/core/internal/traits.d [hunk 3]
```diff
@@ -164,6 +164,75 @@ template isInnerClass(T) if (is(T == class))
...
+        alias Type = X;
+    }
+    else
+    {
+        alias Type = T;
+    }
+}
+
+/// Detect whether type `T` is a static array.
+template isStaticArray(T)
```
Comment (MoonlightSentinel) [inline src/core/internal/traits.d]: Question:
With dlang/dmd#9769 template alias parameters should accept all types and hence this overload would not be required. But removing it results in the following error message:

```
traits.d(196): Error: template instance `isStaticArray!(int[3])` does not match template declaration isStaticArray(alias T)
traits.d(196):        while evaluating: static assert(isStaticArray!(int[3]))
```

Comment (JinShil) [reply src/core/internal/traits.d]: I don't know. @FeepingCreature ?

Comment (FeepingCreature) [reply src/core/internal/traits.d]: Works with run.dlang.org on dmd-nightly... https://run.dlang.io/is/DYSMCo

Are you sure you're using current dmd?

Comment (MoonlightSentinel) [reply src/core/internal/traits.d]: @FeepingCreature My bad, i was using 2.086.1.

**Comment by JinShil** [comment]
I believe I've addressed all new issues. This PR is ready for additional scrutiny.

**Comment by jacob-carlborg** [comment]
Are we sure it's better to not use `__traits(isStaticArray)`? The library implementation will cause two template instantiations. Is that slower?

**Comment by JinShil** [comment]
> Are we sure it's better to not use __traits(isStaticArray)?

I am of the opinion that anything that can be implemented in the library instead of the language should be.  I'm not sure if that's a very popular opinion, a consensus of D contributors, or a policy of the D Language Foundation.

> Is that slower?

Probably, but I doubt that anyone will notice especially given how infrequently it's used.  Also, in my opinion, any performance issues at compile-time are first and foremost a problem with the compiler implementation, not the language.
